chainstate/src/detail/chainstateref/consistency_checker.rs

```diff
@@ -0,0 +1,221 @@
+// Copyright (c) 2021-2024 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::{BTreeMap, BTreeSet};
+
+use chainstate_storage::BlockchainStorageRead;
+use chainstate_types::BlockIndex;
+use common::{
+    chain::{Block, ChainConfig, GenBlock},
+    primitives::{BlockHeight, Id},
+};
+use itertools::{EitherOrBoth, Itertools};
+use logging::log;
+
+use super::calc_min_height_with_allowed_reorg;
+
+// Certain tests check for this panic message.
+const PANIC_MSG: &str = "Inconsistent chainstate";
+
+pub struct ConsistencyChecker<'a, 'b, DbTx> {
+    _db_tx: &'a DbTx,
+    chain_config: &'b ChainConfig,
```

Comment:
Just use an Arc instead of complicating this with another lifetime...

Reply:
> Just use an Arc instead of complicating this with another lifetime...

Actually, separate lifetimes are not needed here, so I've used `'a` for both of the references.